File: benchmarks/bench_data.js
#!/usr/bin/env node
// Node counterpart of bench_data.strada (identical workload).
const fs = require("fs");
const readline = require("readline");

async function main() {
    const path = "/tmp/node_bench_data.csv";
    const rows = 500000;
    const t0 = Date.now();

    let seed = 42;
    const parts = [];
    for (let i = 0; i < rows; i++) {
        seed = (seed * 1103515245 + 12345) % 2147483648;
        parts.push(`2026-06-12T01:00:00,user${seed % 5000},action${Math.floor(seed/7) % 6},${seed % 50000},ok\n`);
    }
    fs.writeFileSync(path, parts.join(""));
    const t1 = Date.now();
    console.log("generate:", rows, (t1 - t0) / 1000);

    const byUser = {}, byAction = {}, bytesByUser = {};
    let parsed = 0;
    const rl = readline.createInterface({ input: fs.createReadStream(path), crlfDelay: Infinity });
    for await (const line of rl) {
        const f = line.split(",");
        if (f.length < 5) continue;
        byUser[f[1]] = (byUser[f[1]] || 0) + 1;
        byAction[f[2]] = (byAction[f[2]] || 0) + 1;
        bytesByUser[f[1]] = (bytesByUser[f[1]] || 0) + (+f[3]);
        parsed++;
    }
    const t2 = Date.now();
    console.log("aggregate:", parsed, "users=" + Object.keys(byUser).length, (t2 - t1) / 1000);

    const ranked = Object.keys(bytesByUser).sort((a, b) => bytesByUser[b] - bytesByUser[a]);
    let report = "";
    for (let i = 0; i < 20; i++) {
        const u = ranked[i];
        report += `${u} events=${byUser[u]} bytes=${bytesByUser[u]}\n`;
    }
    const t3 = Date.now();
    console.log("report:", report.length, (t3 - t2) / 1000);
    fs.unlinkSync(path);
    console.log("total:", (t3 - t0) / 1000);
}
main();

File: benchmarks/bench_data.pl
#!/usr/bin/env perl
# Perl counterpart of bench_data.strada (identical workload).
use strict; use warnings; use Time::HiRes qw(time);

my $path = "/tmp/perl_bench_data.csv";
my $rows = 500000;
my $t0 = time;

my $seed = 42; my $out = "";
for my $i (0..$rows-1) {
    $seed = ($seed * 1103515245 + 12345) % 2147483648;
    my ($user, $action, $bytes) = ($seed % 5000, int($seed/7) % 6, $seed % 50000);
    $out .= "2026-06-12T01:00:00,user$user,action$action,$bytes,ok\n";
}
open my $wfh, '>', $path or die; print $wfh $out; close $wfh;
my $t1 = time;
printf "generate: %d %.6f\n", $rows, $t1 - $t0;

my (%by_user, %by_action, %bytes_by_user);
my $parsed = 0;
open my $fh, '<', $path or die;
while (my $line = <$fh>) {
    chomp $line;
    my @f = split /,/, $line;
    next if @f < 5;
    $by_user{$f[1]}++; $by_action{$f[2]}++; $bytes_by_user{$f[1]} += $f[3];
    $parsed++;
}
close $fh;
my $t2 = time;
printf "aggregate: %d users=%d %.6f\n", $parsed, scalar(keys %by_user), $t2 - $t1;

my @ranked = sort { $bytes_by_user{$b} <=> $bytes_by_user{$a} } keys %bytes_by_user;
my $report = "";
for my $i (0..19) {
    my $u = $ranked[$i];
    $report .= "$u events=$by_user{$u} bytes=$bytes_by_user{$u}\n";
}
my $t3 = time;
printf "report: %d %.6f\n", length($report), $t3 - $t2;
unlink $path;
printf "total: %.6f\n", $t3 - $t0;
